assets/src/loader.rs

```diff
@@ -0,0 +1,348 @@
+    /// given path.
+    pub fn new<P>(alloc: &Allocator, directory: P, pool: Arc<ThreadPool>) -> Self
+        where P: Into<PathBuf>
+    {
+        Loader {
+            contexts: Default::default(),
+            directory: Arc::new(Box::new(Directory::new(alloc, directory))),
+            pool: pool,
+            stores: Default::default(),
+        }
+    }
+
+    pub fn add_store<I, S>(&mut self, name: I, store: S)
+        where I: Into<String>,
+              S: Store + Send + Sync + 'static
+    {
+        self.stores.insert(name.into(), Arc::new(Box::new(store)));
+    }
+
+    /// Registers an asset and inserts a context into the map.
+    pub fn register<A>(&mut self, context: A::Context)
+        where A: Asset + 'static,
+              A::Context: 'static
+    {
+        self.contexts
+            .insert(TypeId::of::<A>(), Box::new(Arc::new(context)));
+    }
+
+    /// Like `load_from`, but doesn't ask the cache for a
+    pub fn reload<A, F, N, S>(&self,
+                              name: N,
+                              format: F,
+                              store: &S)
+                              -> AssetFuture<A, AssetError<A::Error, F::Error, BoxedErr>>
+        where A: Asset + Send + 'static,
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              N: Into<String>,
+              S: Eq + Hash + ?Sized,
+              String: Borrow<S>
+    {
+        let context = self.context::<A>();
+
+        reload_asset_future::<A, F, N, _>(context.clone(),
+                                          format,
+                                          name,
+                                          self.store(store),
+                                          &*self.pool)
+    }
+
+    /// Loads an asset with a given format from the default (directory) store.
+    /// If you want to load from a custom source instead, use `load_from`.
+    ///
+    /// The actual work is done on a worker thread, thus this method immediately returns
+    /// a future.
+    pub fn load<A, F, N>(&self,
+                         id: N,
+                         format: F)
+                         -> AssetFuture<A, AssetError<A::Error, F::Error, BoxedErr>>
+        where A: Asset + Send + 'static,
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              N: Into<String>
+    {
+        self.load_from::<A, F, _, _>(id, format, "")
+    }
+
+    /// Loads an asset with a given id and format from a custom store.
+    /// The actual work is done on a worker thread, thus this method immediately returns
+    /// a future.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the asset wasn't registered.
```

Comment:
You register every type that implements `Asset` (in case you want to load it at some point).